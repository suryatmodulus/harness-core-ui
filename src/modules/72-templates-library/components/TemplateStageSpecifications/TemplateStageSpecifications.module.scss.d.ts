/* eslint-disable */
// this is an auto-generated file
declare const styles: {
  readonly serviceOverrides: string
  readonly contentSection: string
  readonly tabHeading: string
  readonly tabSubHeading: string
  readonly sectionCard: string
  readonly templateBar: string
  readonly inputsContainer: string
  readonly error: string
  readonly stageCard: string
}
export default styles
