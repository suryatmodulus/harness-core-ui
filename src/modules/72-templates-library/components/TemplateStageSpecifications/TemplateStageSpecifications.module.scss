/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.serviceOverrides {
  .content-section {
    padding: var(--spacing-xxlarge) var(--spacing-huge) var(--spacing-13) var(--spacing-huge) !important;
    > .tabHeading {
      font-size: 16px;
      color: var(--black);
      margin-top: var(--spacing-7);
      font-weight: 600;
      line-height: 24px;
      &:first-child {
        margin-top: 0;
      }
    }
    .tabSubHeading {
      font-size: var(--font-size-normal);
      color: var(--grey-700);
      margin-bottom: var(--spacing-5);
      font-weight: 600;
    }
    .section-card {
      width: 1024px;
      display: block;
      padding: var(--spacing-7);
      border-top-left-radius: 0px;
      border-top-right-radius: 0px;
      :global(.InputWithIdentifier--txtNameContainer) {
        max-width: 400px;
        margin-bottom: 0px;
      }
    }
    .templateBar {
      margin-right: 0px !important;
      margin-left: 0px !important;
      width: 1024px;
    }
    .inputs-container {
      position: relative;
      min-height: 300px;
      width: 1024px;
      .error {
        min-height: 200px;
      }
      .stageCard {
        margin: 0px 0px var(--spacing-xxlarge) 0px;
        padding: var(--spacing-5) 0px;
        background: var(--white);
        border: 1px solid transparent;
        box-shadow: var(--elevation-2) !important;
      }
    }
  }
}
