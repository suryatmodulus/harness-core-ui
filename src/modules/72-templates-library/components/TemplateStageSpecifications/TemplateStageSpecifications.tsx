import React, { useContext } from 'react'
import { debounce, defaultTo, isEqual, merge, noop, omit, set } from 'lodash-es'
import { Card, Color, Container, Formik, FormikForm, Heading, Layout, PageError } from '@wings-software/uicore'
import * as Yup from 'yup'
import { useParams } from 'react-router-dom'
import { parse } from 'yaml'
import type { FormikProps } from 'formik'
import produce from 'immer'
import { usePipelineContext } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import type { Error, StageElementConfig } from 'services/cd-ng'
import { useStrings } from 'framework/strings'
import { IdentifierSchema, NameSchema } from '@common/utils/Validation'
import { PageSpinner } from '@common/components'
import { Scope } from '@common/interfaces/SecretsInterface'
import { getIdentifierFromValue, getScopeFromValue } from '@common/components/EntityReference/EntityReference'
import type { ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import { NameId } from '@common/components/NameIdDescriptionTags/NameIdDescriptionTags'
import { useToaster } from '@common/exports'
import { useGetTemplateInputSetYaml } from 'services/template-ng'
import { StepViewType } from '@pipeline/components/AbstractSteps/Step'
import { StageForm } from '@pipeline/components/PipelineInputSetForm/PipelineInputSetForm'
import { StageErrorContext } from '@pipeline/context/StageErrorContext'
import { TemplateTabs } from '@templates-library/components/TemplateStageSetupShell/TemplateStageSetupShellUtils'
import { validateStage } from '@pipeline/components/PipelineStudio/StepUtil'
import { useGlobalEventListener } from '@common/hooks'
import ErrorsStripBinded from '@pipeline/components/ErrorsStrip/ErrorsStripBinded'
import { useStageTemplateActions } from '@pipeline/utils/useStageTemplateActions'
import { TemplateBar } from '@pipeline/components/PipelineStudio/TemplateBar/TemplateBar'
import { setTemplateInputs, TEMPLATE_INPUT_PATH } from '@pipeline/utils/templateUtils'
import css from './TemplateStageSpecifications.module.scss'

declare global {
  interface WindowEventMap {
    SAVE_PIPELINE_CLICKED: CustomEvent<string>
  }
}

export interface TemplateStageValues extends StageElementConfig {
  inputSetTemplate?: StageElementConfig
}

export const TemplateStageSpecifications = (): JSX.Element => {
  const {
    state: {
      selectionState: { selectedStageId = '' }
    },
    allowableTypes,
    updateStage,
    isReadonly,
    getStageFromPipeline
  } = usePipelineContext()
  const { stage } = getStageFromPipeline(selectedStageId)
  const { accountId, orgIdentifier, projectIdentifier } = useParams<ProjectPathProps>()
  const [formValues, setFormValues] = React.useState<TemplateStageValues>(stage?.stage as TemplateStageValues)
  const scope = getScopeFromValue(defaultTo(stage?.stage?.template?.templateRef, ''))
  const { subscribeForm, unSubscribeForm } = React.useContext(StageErrorContext)
  const formikRef = React.useRef<FormikProps<unknown> | null>(null)
  const { submitFormsForTab } = useContext(StageErrorContext)
  const { showError } = useToaster()
  const { getString } = useStrings()

  const onChange = React.useCallback(
    debounce(async (values: StageElementConfig): Promise<void> => {
      await updateStage({ ...stage?.stage, ...values })
    }, 300),
    [stage?.stage, updateStage]
  )

  const {
    data: templateInputYaml,
    error: inputSetError,
    refetch,
    loading
  } = useGetTemplateInputSetYaml({
    templateIdentifier: getIdentifierFromValue(defaultTo(stage?.stage?.template?.templateRef, '')),
    queryParams: {
      accountIdentifier: accountId,
      projectIdentifier: scope === Scope.PROJECT ? projectIdentifier : undefined,
      orgIdentifier: scope === Scope.PROJECT || scope === Scope.ORG ? orgIdentifier : undefined,
      versionLabel: defaultTo(stage?.stage?.template?.versionLabel, '')
    }
  })

  React.useEffect(() => {
    if (!loading && stage?.stage) {
      try {
        const templateInputs = parse(defaultTo(templateInputYaml?.data, ''))
        const mergedTemplateInputs = merge({}, templateInputs, stage?.stage.template?.templateInputs)
        setFormValues(
          produce(stage?.stage as TemplateStageValues, draft => {
            setTemplateInputs(draft, mergedTemplateInputs)
            if (templateInputs) {
              draft.inputSetTemplate = templateInputs
            }
          })
        )
        setTemplateInputs(stage.stage, mergedTemplateInputs)
        updateStage(stage.stage)
      } catch (error) {
        showError(error.message, undefined, 'template.parse.inputSet.error')
      }
    }
  }, [templateInputYaml?.data, loading])

  React.useEffect(() => {
    subscribeForm({ tab: TemplateTabs.OVERVIEW, form: formikRef })
    return () => unSubscribeForm({ tab: TemplateTabs.OVERVIEW, form: formikRef })
  }, [subscribeForm, unSubscribeForm, formikRef])

  useGlobalEventListener('SAVE_PIPELINE_CLICKED', _event => {
    submitFormsForTab(TemplateTabs.OVERVIEW)
  })

  const validateForm = (values: TemplateStageValues) => {
    if (
      isEqual(values.template?.templateRef, stage?.stage?.template?.templateRef) &&
      isEqual(values.template?.versionLabel, stage?.stage?.template?.versionLabel)
    ) {
      onChange?.(omit(values, 'inputSetTemplate'))
      const errorsResponse = validateStage({
        stage: values.template?.templateInputs as StageElementConfig,
        template: values.inputSetTemplate,
        originalStage: stage?.stage?.template?.templateInputs as StageElementConfig,
        getString,
        viewType: StepViewType.DeploymentForm
      })
      return set({}, TEMPLATE_INPUT_PATH, errorsResponse)
    } else {
      return {}
    }
  }

  const { onRemoveTemplate, onOpenTemplateSelector } = useStageTemplateActions()

  return (
    <Container className={css.serviceOverrides} height={'100%'} background={Color.FORM_BG}>
      <ErrorsStripBinded />
      <Layout.Vertical spacing={'xlarge'} className={css.contentSection}>
        {stage?.stage?.template && (
          <TemplateBar
            templateLinkConfig={stage?.stage.template}
            onRemoveTemplate={onRemoveTemplate}
            onOpenTemplateSelector={onOpenTemplateSelector}
            className={css.templateBar}
          />
        )}
        <Formik<TemplateStageValues>
          initialValues={formValues}
          formName="templateStageOverview"
          onSubmit={noop}
          validate={validateForm}
          validationSchema={Yup.object().shape({
            name: NameSchema({
              requiredErrorMsg: getString('pipelineSteps.build.create.stageNameRequiredError')
            }),
            identifier: IdentifierSchema()
          })}
          enableReinitialize={true}
        >
          {(formik: FormikProps<TemplateStageValues>) => {
            window.dispatchEvent(new CustomEvent('UPDATE_ERRORS_STRIP', { detail: TemplateTabs.OVERVIEW }))
            formikRef.current = formik
            return (
              <FormikForm>
                <Card className={css.sectionCard}>
                  <NameId
                    identifierProps={{
                      inputLabel: getString('stageNameLabel'),
                      isIdentifierEditable: false,
                      inputGroupProps: { disabled: isReadonly }
                    }}
                    inputGroupProps={{ placeholder: getString('common.namePlaceholder') }}
                  />
                </Card>
                <Container className={css.inputsContainer}>
                  {loading && <PageSpinner />}
                  {!loading && inputSetError && (
                    <Container height={300}>
                      <PageError
                        message={defaultTo((inputSetError.data as Error)?.message, inputSetError.message)}
                        onClick={() => refetch()}
                      />
                    </Container>
                  )}
                  {!loading && !inputSetError && formik.values.inputSetTemplate && (
                    <Layout.Vertical
                      margin={{ top: 'medium' }}
                      padding={{ top: 'large', bottom: 'large' }}
                      spacing={'large'}
                    >
                      <Heading level={5} color={Color.BLACK}>
                        {getString('templatesLibrary.templateInputs')}
                      </Heading>
                      <StageForm
                        key={`${formik.values.template?.templateRef}-${formik.values.template?.versionLabel || ''}`}
                        template={{ stage: formik.values.inputSetTemplate }}
                        allValues={{ stage: formik.values.template?.templateInputs as StageElementConfig }}
                        path={TEMPLATE_INPUT_PATH}
                        readonly={isReadonly}
                        viewType={StepViewType.InputSet}
                        hideTitle={true}
                        stageClassName={css.stageCard}
                        allowableTypes={allowableTypes}
                      />
                    </Layout.Vertical>
                  )}
                </Container>
              </FormikForm>
            )
          }}
        </Formik>
      </Layout.Vertical>
    </Container>
  )
}
