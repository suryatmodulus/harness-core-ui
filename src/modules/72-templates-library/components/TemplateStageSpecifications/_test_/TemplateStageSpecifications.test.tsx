/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { set } from 'lodash-es'
import produce from 'immer'
import routes from '@common/RouteDefinitions'
import { accountPathProps, pipelineModuleParams, pipelinePathProps } from '@common/utils/routeUtils'
import { PipelineContext } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import pipelineContextMock from '@pipeline/components/PipelineStudio/RightDrawer/__tests__/stateMock'
import { TestWrapper } from '@common/utils/testUtils'
import {
  DefaultNewStageId,
  DefaultNewStageName
} from '@templates-library/components/TemplateStudio/StageTemplateCanvas/StageTemplateForm/StageTemplateForm'
import { TemplateStageSpecifications } from '@templates-library/components/TemplateStageSpecifications/TemplateStageSpecifications'
import { stageMockTemplatesInputYaml } from '@templates-library/TemplatesTestHelper'

window.IntersectionObserver = jest.fn().mockImplementation(() => ({
  observe: () => null,
  unobserve: () => null
}))

jest.mock('framework/strings', () => ({
  ...(jest.requireActual('framework/strings') as any),
  useStrings: jest.fn().mockReturnValue({
    getString: jest.fn().mockImplementation(val => val)
  })
}))

jest.mock('services/template-ng', () => ({
  ...(jest.requireActual('services/template-ng') as any),
  useGetTemplateInputSetYaml: jest
    .fn()
    .mockImplementation(() => ({ data: stageMockTemplatesInputYaml, refetch: jest.fn(), error: null, loading: false })),
  useGetTemplate: jest.fn().mockImplementation(() => ({
    status: 'SUCCESS',
    data: {
      data: {
        name: 'New Stage Name',
        identifier: 'new_stage_name',
        versionLabel: 'v1'
      }
    }
  }))
}))

jest.mock('services/cd-ng', () => ({
  useGetServiceAccessList: jest.fn().mockImplementation(() => ({
    loading: false,
    data: {
      status: 'SUCCESS',
      data: []
    },
    refetch: jest.fn()
  }))
}))

describe('<TemplateStageSpecifications /> tests', () => {
  test('snapshot test', async () => {
    const templateStage = {
      name: DefaultNewStageName,
      identifier: DefaultNewStageId,
      template: {
        templateRef: 'new_stage_name',
        versionLabel: 'v1',
        templateInputs: {
          type: 'Deployment',
          spec: {
            serviceConfig: {
              serviceRef: '<+input>'
            },
            infrastructure: {
              infrastructureDefinition: {
                type: 'KubernetesDirect',
                spec: {
                  namespace: '<+input>'
                }
              }
            },
            execution: {
              steps: [
                {
                  step: {
                    identifier: 'Step_1',
                    type: 'ShellScript',
                    spec: {
                      source: {
                        type: 'Inline',
                        spec: {
                          script: '<+input>'
                        }
                      }
                    }
                  }
                }
              ]
            }
          }
        }
      }
    }
    const context = produce(pipelineContextMock, draft => {
      delete draft.state.pipeline.stages
      set(draft, 'state.pipeline.stages[0].stage', templateStage)
      set(draft, 'state.selectionState', {
        selectedStageId: templateStage.identifier
      })
      draft.getStageFromPipeline = jest.fn().mockReturnValue({
        stage: {
          stage: templateStage
        }
      })
    })
    const { container } = render(
      <PipelineContext.Provider value={context}>
        <TestWrapper
          path={routes.toPipelineStudio({ ...accountPathProps, ...pipelinePathProps, ...pipelineModuleParams })}
          pathParams={{
            pipelineIdentifier: 'stage1',
            accountId: 'accountId',
            projectIdentifier: 'Milos2',
            orgIdentifier: 'CV',
            module: 'cd'
          }}
        >
          <TemplateStageSpecifications />
        </TestWrapper>
      </PipelineContext.Provider>
    )
    expect(container).toMatchSnapshot()
  })
})
