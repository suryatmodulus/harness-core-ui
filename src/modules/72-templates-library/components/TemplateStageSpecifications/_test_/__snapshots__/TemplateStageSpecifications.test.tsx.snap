// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TemplateStageSpecifications /> tests snapshot test 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main serviceOverrides StyledProps--background-form"
    style="height: 100%;"
  >
    <div
      class="Layout--vertical Layout--layout-spacing-xlarge StyledProps--main contentSection"
    >
      <div
        class="StyledProps--font StyledProps--main container templateBar StyledProps--padding-top-small StyledProps--padding-right-medium StyledProps--padding-bottom-small StyledProps--padding-left-medium StyledProps--background-primary6 StyledProps--border"
      >
        <div
          class="Layout--horizontal Layout--layout-spacing-small StyledProps--main StyledProps--flex StyledProps--flex-alignItems-center StyledProps--flex-justifyContent-flex-start"
        >
          <span
            class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-white"
            data-icon="template-library"
          >
            <svg
              height="11"
              viewBox="0 0 12 13"
              width="11"
              xmlns="http://www.w3.org/2000/svg"
              xmlns:xlink="http://www.w3.org/1999/xlink"
            >
              <defs>
                <path
                  d="M1.8 13c.2 0 .4-.3.4-.6V.6C2.2.3 2 0 1.8 0H.4C.2 0 0 .3 0 .6v11.7c0 .4.2.7.4.7h1.4zm8.8 0l.9-.3c.3-.1.5-.4.4-.8L8.9.6C8.8.3 8.5.1 8.1.2l-.9.3c-.4 0-.5.4-.5.7l3.1 11.3c.1.4.5.6.8.5zm-5 0c.2 0 .4-.3.4-.7V.6c0-.3-.2-.6-.4-.6H4.2c-.2 0-.4.3-.4.6v11.7c0 .4.2.7.4.7h1.4z"
                  id="template-library_svg__a"
                />
              </defs>
              <clippath
                id="template-library_svg__b"
              >
                <use
                  overflow="visible"
                  xlink:href="#template-library_svg__a"
                />
              </clippath>
              <path
                clip-path="url(#template-library_svg__b)"
                d="M-5-5h22v23H-5z"
                fill="#25a6f7"
              />
            </svg>
          </span>
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-size-small StyledProps--color StyledProps--color-white"
            style="flex-grow: 1;"
          >
            Using Template: New Stage Name (v1)
          </p>
          <span
            class="bp3-popover-wrapper main"
          >
            <span
              class="bp3-popover-target"
            >
              <span
                class="bp3-icon bp3-icon-more StyledProps--main actionButton StyledProps--color StyledProps--color-white"
                icon="more"
              >
                <svg
                  data-icon="more"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    more
                  </desc>
                  <path
                    d="M2 6.03a2 2 0 100 4 2 2 0 100-4zM14 6.03a2 2 0 100 4 2 2 0 100-4zM8 6.03a2 2 0 100 4 2 2 0 100-4z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </span>
          </span>
        </div>
      </div>
      <div
        class="OverlaySpinner--overlaySpinner"
      >
        <form
          class="FormikForm--main"
        >
          <div>
            <div
              class="bp3-card bp3-elevation-0 Card--card sectionCard"
            >
              <div
                class="InputWithIdentifier--txtNameContainer TextInput--main"
              >
                <div
                  class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main InputWithIdentifier--txtIdContainer"
                >
                  <div
                    class="Text--withDocsTooltip"
                    data-tooltip-id="templateStageOverview_identifier"
                  >
                    <p
                      class="StyledProps--font StyledProps--main"
                    >
                      ID
                    </p>
                  </div>
                  <p
                    class="StyledProps--font StyledProps--main"
                  >
                    :
                  </p>
                  <span
                    class="bp3-popover-wrapper"
                  >
                    <span
                      class="bp3-popover-target"
                    >
                      <div
                        class="bp3-editable-text bp3-disabled InputWithIdentifier--idValue"
                      >
                        <span
                          class="bp3-editable-text-content"
                          style="height: 1px; line-height: 1px; min-width: 0;"
                        >
                          stage_name
                        </span>
                      </div>
                    </span>
                  </span>
                </div>
                <div
                  class="bp3-form-group"
                  data-id="name-0"
                >
                  <label
                    class="bp3-label"
                    for="name"
                  >
                    <span
                      class="Tooltip--acenter"
                      data-tooltip-id="templateStageOverview_name"
                    >
                      stageNameLabel
                    </span>
                     
                    <span
                      class="bp3-text-muted"
                    />
                  </label>
                  <div
                    class="bp3-form-content"
                  >
                    <div
                      class="bp3-input-group"
                    >
                      <input
                        autocomplete="off"
                        class="bp3-input"
                        name="name"
                        type="text"
                        value="Stage Name"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="StyledProps--font StyledProps--main inputsContainer"
            >
              <div
                class="Layout--vertical Layout--layout-spacing-large StyledProps--main StyledProps--margin-top-medium StyledProps--padding-top-large StyledProps--padding-bottom-large"
              >
                <h5
                  class="StyledProps--font StyledProps--font-h5 StyledProps--main StyledProps--color StyledProps--color-black"
                  level="5"
                >
                  templatesLibrary.templateInputs
                </h5>
                <div
                  id="Stage.undefined"
                >
                  <div
                    class="topAccordion stageCard"
                  >
                    <div
                      class="accordionSummary"
                      id="Stage.undefined.Service"
                    >
                      <div
                        class="inputheader"
                      >
                        service
                      </div>
                      <div
                        class="nestedAccordions"
                      >
                        <div
                          class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
                          style="align-items: center;"
                        >
                          <div
                            class="bp3-form-group inputWidth"
                          >
                            <label
                              class="bp3-label"
                              for="template.templateInputs.spec.serviceConfig.serviceRef"
                            >
                              <span
                                class="Tooltip--acenter"
                                data-tooltip-id="specifyYourService"
                              >
                                cd.pipelineSteps.serviceTab.specifyYourService
                              </span>
                               
                              <span
                                class="bp3-text-muted"
                              />
                            </label>
                            <div
                              class="bp3-form-content"
                            >
                              <div
                                class="Layout--horizontal StyledProps--main MultiTypeInput--main MultiTypeInput--disabled"
                              >
                                <div
                                  class="TextInput--main MultiTypeInput--input"
                                >
                                  <div
                                    class="bp3-input-group bp3-disabled"
                                  >
                                    <input
                                      class="bp3-input"
                                      disabled=""
                                      name="template.templateInputs.spec.serviceConfig.serviceRef"
                                      placeholder="<+input>"
                                      type="text"
                                      value="<+input>"
                                    />
                                  </div>
                                </div>
                                <span
                                  class="bp3-popover-wrapper MultiTypeInput--wrapper"
                                >
                                  <span
                                    class="bp3-popover-target"
                                  >
                                    <button
                                      class="MultiTypeInput--btn MultiTypeInput--RUNTIME"
                                    >
                                      <span
                                        class="bp3-icon StyledProps--main"
                                        data-icon="runtime-input"
                                      >
                                        <svg
                                          height="12"
                                          viewBox="0 0 10 7.97"
                                          width="12"
                                          xmlns="http://www.w3.org/2000/svg"
                                        >
                                          <path
                                            d="M8.14 4L8 3.72a2.53 2.53 0 01-.12-.3l-.16-.35c-.11-.25-.22-.38-.32-.38s-.28.15-.37.44C7 3.1 7 3.07 7 3a.73.73 0 01.24-.46.66.66 0 01.42-.26c.25 0 .5.35.76 1.05l.07.17.07-.12c.47-.73.86-1.1 1.18-1.1a.71.71 0 01.29.07l-.4.42h-.11c-.24 0-.52.28-.86.83l-.09.16.07.2c.29.79.53 1.18.71 1.18s.31-.12.4-.36q.09.06.09.12t-.27.39a.74.74 0 01-.45.23c-.25 0-.51-.36-.77-1.07l-.12-.25-.1.17c-.44.76-.87 1.14-1.28 1.14a.65.65 0 01-.39-.12L6.84 5a.32.32 0 00.23.1c.25 0 .53-.25.83-.74l.17-.28zm-4-.32L1.7.59h2.47S5 .52 5.35 1a3.28 3.28 0 01.42 1.14h.31L5.94 0H0v.31l3.06 3.74L0 7.62V8h6.08l.38-2.36-.31-.07a2.43 2.43 0 01-.52 1c-.42.48-2 .41-2 .41H1.39z"
                                            fill="current-color"
                                          />
                                        </svg>
                                      </span>
                                    </button>
                                  </span>
                                </span>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="accordionSummary"
                      id="Stage.undefined.Infrastructure"
                    >
                      <div
                        class="inputheader"
                      >
                        infrastructureText
                      </div>
                      <div
                        class="nestedAccordions"
                        style="width: 50%;"
                      />
                      <div
                        class="nestedAccordions"
                      >
                        <div
                          class="Layout--vertical Layout--layout-spacing-small StyledProps--main containerPadding StyledProps--padding-medium"
                        >
                          <div
                            class="formGroup md"
                          >
                            <div
                              class="bp3-form-group"
                            >
                              <label
                                class="bp3-label"
                                for="template.templateInputs.spec.infrastructure.infrastructureDefinition.spec.namespace"
                              >
                                <span
                                  class="Tooltip--acenter"
                                  data-tooltip-id="templateStageOverview_template.templateInputs.spec.infrastructure.infrastructureDefinition.spec.namespace"
                                >
                                  common.namespace
                                </span>
                                 
                                <span
                                  class="bp3-text-muted"
                                />
                              </label>
                              <div
                                class="bp3-form-content"
                              >
                                <div
                                  class="Layout--horizontal StyledProps--main MultiTypeInput--main MultiTypeInput--disabled"
                                >
                                  <div
                                    class="TextInput--main MultiTypeInput--input"
                                  >
                                    <div
                                      class="bp3-input-group bp3-disabled"
                                    >
                                      <input
                                        class="bp3-input"
                                        disabled=""
                                        name="template.templateInputs.spec.infrastructure.infrastructureDefinition.spec.namespace"
                                        placeholder="<+input>"
                                        type="text"
                                        value="<+input>"
                                      />
                                    </div>
                                  </div>
                                  <span
                                    class="bp3-popover-wrapper MultiTypeInput--wrapper"
                                  >
                                    <span
                                      class="bp3-popover-target"
                                    >
                                      <button
                                        class="MultiTypeInput--btn MultiTypeInput--RUNTIME"
                                      >
                                        <span
                                          class="bp3-icon StyledProps--main"
                                          data-icon="runtime-input"
                                        >
                                          <svg
                                            height="12"
                                            viewBox="0 0 10 7.97"
                                            width="12"
                                            xmlns="http://www.w3.org/2000/svg"
                                          >
                                            <path
                                              d="M8.14 4L8 3.72a2.53 2.53 0 01-.12-.3l-.16-.35c-.11-.25-.22-.38-.32-.38s-.28.15-.37.44C7 3.1 7 3.07 7 3a.73.73 0 01.24-.46.66.66 0 01.42-.26c.25 0 .5.35.76 1.05l.07.17.07-.12c.47-.73.86-1.1 1.18-1.1a.71.71 0 01.29.07l-.4.42h-.11c-.24 0-.52.28-.86.83l-.09.16.07.2c.29.79.53 1.18.71 1.18s.31-.12.4-.36q.09.06.09.12t-.27.39a.74.74 0 01-.45.23c-.25 0-.51-.36-.77-1.07l-.12-.25-.1.17c-.44.76-.87 1.14-1.28 1.14a.65.65 0 01-.39-.12L6.84 5a.32.32 0 00.23.1c.25 0 .53-.25.83-.74l.17-.28zm-4-.32L1.7.59h2.47S5 .52 5.35 1a3.28 3.28 0 01.42 1.14h.31L5.94 0H0v.31l3.06 3.74L0 7.62V8h6.08l.38-2.36-.31-.07a2.43 2.43 0 01-.52 1c-.42.48-2 .41-2 .41H1.39z"
                                              fill="current-color"
                                            />
                                          </svg>
                                        </span>
                                      </button>
                                    </span>
                                  </span>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="accordionSummary"
                      id="Stage.undefined.Execution"
                    >
                      <div
                        class="inputheader"
                      >
                        executionText
                      </div>
                      <div
                        class="nestedAccordions"
                      >
                        <div
                          class="Layout--vertical Layout--layout-spacing-medium StyledProps--main StyledProps--padding-top-medium"
                        >
                          <label
                            class="Label--main"
                          >
                            <span
                              class="bp3-icon StyledProps--main StyledProps--padding-right-small StyledProps--color StyledProps--color-grey700"
                              data-icon="command-shell-script"
                            >
                              <svg
                                height="16"
                                viewBox="0 0 14 11"
                                width="16"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M4.2 2.8c0-.4-.3-.7-.7-.7-.2 0-.4 0-.5.2L.2 5c-.1.1-.2.3-.2.5s.1.4.2.5L3 8.7c.1.1.3.2.5.2.4 0 .7-.3.7-.7 0-.2-.1-.4-.2-.5L1.7 5.5 4 3.2c.1-.1.2-.3.2-.4zM8.4 0c-.3 0-.6.2-.7.5l-2.8 9.6v.2c0 .4.3.7.7.7.3 0 .6-.2.7-.5L9.1.9V.7c0-.4-.3-.7-.7-.7zm5.4 5L11 2.3c-.1-.1-.3-.2-.5-.2-.4 0-.7.3-.7.7 0 .2.1.4.2.5l2.3 2.3L10 7.8c-.1.1-.2.3-.2.4 0 .4.3.7.7.7.2 0 .4-.1.5-.2L13.8 6c.1-.1.2-.3.2-.5s-.1-.4-.2-.5z"
                                  fill="currentColor"
                                />
                              </svg>
                            </span>
                            pipeline.execution.stepTitlePrefix
                            pipeline.stepLabel
                          </label>
                          <div>
                            <form
                              class="FormikForm--main"
                            >
                              <div>
                                <div
                                  class="formGroup alignStart md"
                                >
                                  <div
                                    class="bp3-form-group formGroup"
                                    data-id="template.templateInputs.spec.execution.steps[0].step.spec.source.spec.script-0"
                                  >
                                    <label
                                      class="bp3-label"
                                      for="template.templateInputs.spec.execution.steps[0].step.spec.source.spec.script"
                                    >
                                      <div
                                        class="formLabel"
                                      >
                                        <span
                                          class="Tooltip--acenter"
                                          data-tooltip-id="templateStageOverview_template.templateInputs.spec.execution.steps[0].step.spec.source.spec.script"
                                        >
                                          script
                                        </span>
                                        <div
                                          class="bp3-popover-wrapper typeSelectorWrapper"
                                        >
                                          <div
                                            class="bp3-popover-target typeSelector"
                                          >
                                            <button
                                              class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main btn Button--iconOnly Button--without-shadow"
                                              type="button"
                                            >
                                              <span
                                                class="bp3-button-text"
                                              >
                                                <span
                                                  class="bp3-icon StyledProps--main icon runtime"
                                                  data-icon="runtime-input"
                                                >
                                                  <svg
                                                    height="12"
                                                    viewBox="0 0 10 7.97"
                                                    width="12"
                                                    xmlns="http://www.w3.org/2000/svg"
                                                  >
                                                    <path
                                                      d="M8.14 4L8 3.72a2.53 2.53 0 01-.12-.3l-.16-.35c-.11-.25-.22-.38-.32-.38s-.28.15-.37.44C7 3.1 7 3.07 7 3a.73.73 0 01.24-.46.66.66 0 01.42-.26c.25 0 .5.35.76 1.05l.07.17.07-.12c.47-.73.86-1.1 1.18-1.1a.71.71 0 01.29.07l-.4.42h-.11c-.24 0-.52.28-.86.83l-.09.16.07.2c.29.79.53 1.18.71 1.18s.31-.12.4-.36q.09.06.09.12t-.27.39a.74.74 0 01-.45.23c-.25 0-.51-.36-.77-1.07l-.12-.25-.1.17c-.44.76-.87 1.14-1.28 1.14a.65.65 0 01-.39-.12L6.84 5a.32.32 0 00.23.1c.25 0 .53-.25.83-.74l.17-.28zm-4-.32L1.7.59h2.47S5 .52 5.35 1a3.28 3.28 0 01.42 1.14h.31L5.94 0H0v.31l3.06 3.74L0 7.62V8h6.08l.38-2.36-.31-.07a2.43 2.43 0 01-.52 1c-.42.48-2 .41-2 .41H1.39z"
                                                      fill="current-color"
                                                    />
                                                  </svg>
                                                </span>
                                              </span>
                                            </button>
                                          </div>
                                        </div>
                                      </div>
                                       
                                      <span
                                        class="bp3-text-muted"
                                      />
                                    </label>
                                    <div
                                      class="bp3-form-content"
                                    >
                                      <div
                                        class="bp3-form-group bp3-disabled runtimeDisabled"
                                        data-id="template.templateInputs.spec.execution.steps[0].step.spec.source.spec.script-1"
                                      >
                                        
                                        <div
                                          class="bp3-form-content"
                                        >
                                          <div
                                            class="bp3-input-group bp3-disabled"
                                          >
                                            <input
                                              autocomplete="off"
                                              class="bp3-input"
                                              disabled=""
                                              name="template.templateInputs.spec.execution.steps[0].step.spec.source.spec.script"
                                              type="text"
                                              value="<+input>"
                                            />
                                          </div>
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </form>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
